L1Trigger/TrackerTFP/plugins/ProducerKFin.cc

```diff
@@ -0,0 +1,207 @@
+        int nStubsSF(0);
+        for (int channel = 0; channel < dataFormats_->numChannel(Process::sf); channel++) {
+          const int index = region * dataFormats_->numChannel(Process::sf) + channel;
+          const TTDTC::Stream& stream = streams[index];
+          nStubsSF += accumulate(stream.begin(), stream.end(), 0, [](int& sum, const TTDTC::Frame& frame){ return sum += frame.first.isNonnull() ? 1 : 0; });
+        }
+        vector<StubSF> stubsSF;
+        stubsSF.reserve(nStubsSF);
+        for (int channel = 0; channel < dataFormats_->numChannel(Process::sf); channel++) {
+          const int index = region * dataFormats_->numChannel(Process::sf) + channel;
+          for (const TTDTC::Frame& frame : streams[index])
+            if (frame.first.isNonnull())
+              stubsSF.emplace_back(frame, dataFormats_);
+        }
+        vector<deque<TTDTC::Frame>> dequesStubs(dataFormats_->numChannel(Process::kf) * setup_->numLayers());
+        vector<deque<FrameTrack>> dequesTracks(dataFormats_->numChannel(Process::kf));
+        int i(0);
+        for (const TTTrack<Ref_Phase2TrackerDigi_>& ttTrack : ttTracks) {
+          if ((int)ttTrack.phiSector() / setup_->numSectorsPhi() != region) {
+            i++;
+            continue;
+          }
+          const int sectorPhi = ttTrack.phiSector() % setup_->numSectorsPhi();
+          deque<FrameTrack>& tracks = dequesTracks[sectorPhi];
+          const int binEta = ttTrack.etaSector();
+          const int binZT = dfZT.toUnsigned(dfZT.integer(ttTrack.z0()));
+          const int binCot = dfCot.toUnsigned(dfCot.integer(ttTrack.tanL()));
+          StubSF* stubSF = nullptr;
+          TTBV hitPattern(0, setup_->numLayers());
+          vector<int> layerCounts(setup_->numLayers(), 0);
+          for (const TTStubRef& ttStubRef : ttTrack.getStubRefs()) {
+            const int layerId = setup_->layerId(ttStubRef);
+            const int layerIdKF = layerEncoding_->layerIdKF(binEta, binZT, binCot, layerId);
+            hitPattern.set(layerIdKF);
+            if (layerCounts[layerIdKF] == setup_->kfMaxStubsPerLayer())
+              continue;
+            layerCounts[layerIdKF]++;
+            deque<TTDTC::Frame>& stubs = dequesStubs[sectorPhi * setup_->numLayers() + layerIdKF];
+            auto identical = [ttStubRef, ttTrack](const StubSF& stub){
+              return (int)ttTrack.trackSeedType() == stub.trackId() && ttStubRef == stub.ttStubRef();
+            };
+            stubSF = &*find_if(stubsSF.begin(), stubsSF.end(), identical);
+            const StubKFin stubKFin(*stubSF, layerIdKF, ttTrack.hitPattern());
+            stubs.emplace_back(stubKFin.frame());
+          }
+          const TTBV& layerMap = setup_->layerMap(hitPattern, layerCounts);
+          const TTBV& maybePattern = layerEncoding_->maybePattern(binEta, binZT, binCot);
+          TrackKFin track(*stubSF, TTTrackRef(handleTTTracks, i++), hitPattern, layerMap, maybePattern);
```

Comment:
I don't understand the TrackKFin constructor. Tracks should have multiple stubs, but the first argument is giving it only a single stub. And the index "i" of TTTrackRef is the number of tracks seen in nonants other than the current one, whereas I would have expected the code to set it to the index of the TTTrack in the TTTrack collection.

Reply:
Tracks in f/w don't have multiple stubs.
There is only one collection of TTTracks not one per nonant. If i would count the tracks per nonant you would create TTTrackRefs from the first nonant instead the correct track.